src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +287,335 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
   f.mk' ((y : M) * x) y = f.1 x :=
 by rw [mul_comm, mk'_mul_cancel_right]
 
+variables {g : M →* P}
+
+@[to_additive] lemma comp_eq_of_eq {T : submonoid P} {Q : Type*} [comm_monoid Q]
+  (hg : ∀ y : S, g y ∈ T) (k : localization_map T Q)
+  {x y} (h : f.1 x = f.1 y) : k.1 (g x) = k.1 (g y) :=
+let ⟨c, hc⟩ := (f.4 _ _).1 h in (k.4 _ _).2
+  ⟨⟨g c, hg c⟩, show g _ * g _ = g _ * g _, by rw [←g.map_mul, hc, g.map_mul]⟩
+
+@[to_additive] lemma eq_of_eq (hg : ∀ y : S, is_unit (g y)) {x y} (h : f.1 x = f.1 y) :
+  g x = g y :=
+begin
+  obtain ⟨c, hc⟩ := (f.4 _ _).1 h,
+  rw [←mul_one (g x), ←is_unit.mul_lift_right_inv (g.restrict S) hg c],
+  show _ * (g _ * _) = _,
+  rw [←mul_assoc, ←g.map_mul, hc, mul_inv_left hg, g.map_mul, mul_comm],
+end
+
+variables (hg : ∀ y : S, is_unit (g y))
+
+/-- Given a localization map `f : M →* N` and a map of `comm_monoid`s `g : M →* P` such that
+    `g y` is invertible for all `y : S`, the homomorphism induced from `N` to `P` sending `z : N`
+    to `g x * (g y)⁻¹`, where `(x, y) : M × S` are such that `z = f x * (f y)⁻¹`. -/
+@[to_additive "Given a localization map `f : M →+ N` and a map of `add_comm_monoid`s `g : M →+ P` such that `g y` is invertible for all `y : S`, the homomorphism induced from `N` to `P` sending `z : N` to `g x - g y`, where `(x, y) : M × S` are such that `z = f x - f y`."]
+noncomputable def lift : N →* P :=
+{ to_fun := λ z, g (sec S f.1 z).1 * ↑(is_unit.lift_right (g.restrict S) hg (sec S f.1 z).2)⁻¹,
+  map_one' := by rw [mul_inv_left, mul_one]; exact f.eq_of_eq hg (by rw [←sec_spec f.3, one_mul]),
+  map_mul' := λ x y, by rw [mul_inv_left hg, ←mul_assoc, ←mul_assoc, mul_inv_right hg,
+    mul_comm _ (g (sec S f.1 y).1), ←mul_assoc, ←mul_assoc, mul_inv_right hg];
+    repeat { rw ←g.map_mul };
+    exact f.eq_of_eq hg (by repeat { rw f.1.map_mul <|> rw sec_spec' f.3 }; ac_refl) }
+
+variables {S g}
+
+@[to_additive] lemma lift_mk' (x y) :
+  f.lift hg (f.mk' x y) = g x * ↑(is_unit.lift_right (g.restrict S) hg y)⁻¹ :=
+(mul_inv hg).2 $ f.eq_of_eq hg $ by
+  rw [f.1.map_mul, f.1.map_mul, sec_spec' f.3, mul_assoc, f.mk'_spec, mul_comm]
+
+@[to_additive] lemma lift_spec (z v) :
+  f.lift hg z = v ↔ g (sec S f.1 z).1 = g (sec S f.1 z).2 * v :=
+mul_inv_left hg _ _ v
+
+@[to_additive] lemma lift_spec_mul (z w v) :
+  f.lift hg z * w = v ↔ g (sec S f.1 z).1 * w = g (sec S f.1 z).2 * v :=
+begin
+  rw mul_comm,
+  show _ * (_ * _) = _ ↔ _,
+  rw [←mul_assoc, mul_inv_left hg, mul_comm],
+end
+
+@[to_additive] lemma lift_mk'_spec (x v) (y : S) :
+  f.lift hg (f.mk' x y) = v ↔ g x = g y * v :=
+by rw f.lift_mk' hg; exact mul_inv_left hg _ _ _
+
+@[to_additive] lemma lift_mul_right (z) :
+  f.lift hg z * g (sec S f.1 z).2 = g (sec S f.1 z).1 :=
+show _ * _ * _ = _, by erw [mul_assoc, is_unit.lift_right_inv_mul, mul_one]
+
+@[to_additive] lemma lift_mul_left (z) :
+  g (sec S f.1 z).2 * f.lift hg z = g (sec S f.1 z).1 :=
+by rw [mul_comm, lift_mul_right]
+
+@[to_additive] lemma lift_of (x : M) :
```

Comment:
This should be `@[simp]`. Also, is there a better name than `lift_of`, seeing as we're not using the name `of` any more.

Reply:
I've made it ```lift_eq``` for now, which sucks. I considered ```lift_on_beta``` but that'd be more appropriate for ```lift_mk'```. I also considered ```lift_mk'_one``` or ```lift_localization_map``` or variations on the last one. Do you have a preference?